### ISSUE TITLE
Feature Request: YAML Front Matter extension

### ISSUE THREAD

**Issue by colinodell** [thephpleague/commonmark#442]
[YAML front matter](https://jekyllrb.com/docs/front-matter/) is commonly used to provide additional metadata about a Markdown file.  Why not include that functionality here as an optional extension?  We could expose the parsed data through the root `Document` node's `$data` array.

Taking this a step further - if the front matter contains a `configuration` key, merge its contents into the current environment's configuration, so that you can adjust the parser's behavior on-the-fly on a document-by-document basis?  This could be really powerful!

Imagine we implement the Table of Contents extension in #441, and you have a Markdown file like this:

```md
---
title: This Is My Awesome Blog Post
configuration:
    table_of_contents:
        max_level: 2
---

# Some heading with level 1

## Some heading with level 2

### Some heading with level 3
```

The `configuration` values in the YAML front-matter would automagically tell the Table of Contents extension to only include the first two headers in the ToC.

(For security reasons, that `configuration` override behavior would be disabled by default, but we'd expose a simple option to enable it when instantiating your converter/environment and converting inputs that you trust).

**Comment by colinodell** [comment]
Until #447 is resolved, being able to override configuration settings per-document could lead to some nasty side effects, so I'm punting this feature to 2.0.

**Comment by markcarver** [comment]
Oooh. Me likey 😃 

I actually spent quite a bit of time implementing this in Drupal core as a standalone component, the latest patch is here: https://www.drupal.org/project/drupal/issues/3064854#comment-13339277

Despite the bikeshedding and political partisanship against front-end work, it's quite stable and thoroughly tested.

p.s. a lot of work has gone into the markdown module to finally provide the proper APIs and UI for extensions: https://www.drupal.org/project/markdown/issues/2952435#comment-13580075

> For security reasons, that `configuration` override behavior would be disabled by default, but we'd expose a simple option to enable it when instantiating your converter/environment and converting inputs that you trust

Do you mean something like Drupal's schema types? Where we can specify which parameters are allowed and what types they should be? i.e.

```yaml
title
  type: string
configuration:
  type: mapping
  mapping:
    table_of_contents:
      type: mapping
      mapping:
        max_level:
          type: integer
```

If so, that would be fantastic as it should be relatively easy to convert what we're already doing:
https://git.drupalcode.org/project/markdown/-/blob/26fdcbfdcca58961d265de32614962f7fac9cb7a/config/schema/markdown.schema.yml#L45-73

**Comment by hason** [comment]
@colinodell Maybe it would be good to also support JSON or TOML format. See https://github.com/webuni/front-matter/blob/master/src/FrontMatter.php

**Comment by markcarver** [comment]
I'm not a fan of allowing JSON or TOML.

TL;DR from  https://www.drupal.org/project/drupal/issues/3064854:

1. Front Matter is/was only ever intended to be YAML, not JSON nor TOML.
2. Introducing more than one way to do a single task only introduces inherent confusion on which one to use in the first place.
3. Increases difficulty in providing support or c&p examples from one language to another.

I'm not saying that `webuni/front-matter` still couldn't be used, but if it is, it should likely be limited to just YAML.

**Comment by colinodell** [comment]
AFAIK, [Hugo](https://gohugo.io/content-management/front-matter/) is the only major tool supporting JSON or TOML.

I see us likely implementing something to what you have @hason where's there's a generic interface for decoding the front matter, but we'll likely only include a single YAML-based implementation.  (Although, json would be super easy to include too...  But we probably wouldn't do TOML)

**Comment by onigoetz** [comment]
I see this has been implemented in #497, does this issue still need to be open ?

Also quick question on the implementation, the only way to read the front-matter is to render the document if I understand it correctly ?

Since I use commonmark in a tool to create documentation from a folder of markdown files (Daux.io) I give the possibility to give the pages a custom title by adding `title:` in the front-matter, this title is used to generate the navigation, thus, when generating a single page, we need the full navigation. And it would be too costly to generate all pages to just get the front-matter for that single page.

I just point this to give a different perspective, I understand my use case might be an edge-case.

**Comment by colinodell** [comment]
I don't know why I closed that PR without merging it... :detective:

**Comment by colinodell** [comment]
Looks like Github might have auto-closed it based on us renaming the main branch? The timestamps seem to match up: https://github.com/thephpleague/commonmark/issues/502  Anyway, I've reopened it as #532 

I'm not opposed to adding a standalone class that can parse front-matter ad-hoc - I can certainly see the value there.  I'll add a note to myself in the PR to include that before merging :)

**Comment by colinodell** [comment]
@onigoetz Would something like 7b8637cb2752bf547a95e51207e9f479f3e85304 work for you?

**Comment by onigoetz** [comment]
Yes, that would work for me, thanks for taking the time on this :)

**Comment by colinodell** [comment]
After going down a _very deep_ rabbit hole, I've come to the realization that allowing the extension to configure the parser based on `configuration` values in frontmatter may not be feasible due to a chicken-and-egg problem.

A single `Environment` instance should contain the same set of extensions/parsers/renderers/etc each time Markdown is parsed.  We could allow configuration to change in between parsings, but [some](https://github.com/thephpleague/commonmark/blob/latest/src/Extension/Mention/MentionExtension.php#L24) [extensions](https://github.com/thephpleague/commonmark/blob/latest/src/Extension/InlinesOnly/InlinesOnlyExtension.php#L53) rely on configuration during registration to determine if certain parsers/renderers/etc should be added.  But if we allow configuration to vary between parsings based then we either have to:

1. Remove the ability for extensions for vary what they register based on configuration (not ideal); or,
2. Allow the shared, multi-use `Environment` to be changed during parsings which will affect future parsings (deal breaker); or,
3. Stop sharing a single `Environment` instance between multiple parsings, maybe clone it each time instead? (not feasible)

The third option sounds the most promising.  Instead of reusing the `Environment`, we could use some kind of `EnvironmentBuilder` which lazily instantiates a new `Environment` for each document being parsed.  But due to the same issues behind `#1`, we'd have to instantiate the extensions so the Front Matter extension can read the embedded configuration, and now we're back to square one :confused:

So, given the current architecture, I don't think this load-configs-from-frontmatter-via-extension feature is feasible for v2.0.  Nevertheless, if somebody wants this functionality for themselves, they could easily implement it by writing code similar to [my comment above](https://github.com/thephpleague/commonmark/issues/442#issuecomment-687806047) and passing the parsed configuration into the parser before executing it.  We just can't automagically bake that functionality into an extension.

I'm therefore going to mark this feature request as closed since everything else has been implemented :)
